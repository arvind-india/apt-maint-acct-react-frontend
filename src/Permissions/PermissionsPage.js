import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { Table } from 'reactstrap'
import { Router } from 'react-router-dom'

import {
  MdAdd,
  MdVisibility,
  MdEdit,
  MdDelete
} from 'react-icons/lib/md' // material design icons

import {
  Button
} from 'reactstrap'

import { history } from '../_helpers'
import { PrivateRoute, FlashMessage } from '../_components'
import { permissionActions as actions } from '../_actions'
import { default as detailsPage } from './PermissionDetailsPage'

let url = '/permissions'
let module = 'permissions'

export class Permissions extends React.Component {

  constructor(props) {
    super(props)
    this.handleDeleteModel = this.handleDeleteModel.bind(this)
  }

  componentDidMount() {
    //this.props.dispatch(actions.getAll())
    this.props.getAll()
  }

  render() {
    //console.log('Props in PermissionsPage: ', this.props)
    const { permissions, alert, authzn, trackHistory } = this.props
    let models = permissions
    let hist = trackHistory?history:{}
    return (
      <div>
        <h3>Permissions List</h3>
        {alert.message && <FlashMessage text={alert.message} color={alert.color} delay={2100}/>}
        {models.loading && <em>Loading models...}</em>}
        {models.error && <span className="text-danger">ERROR: {models.error}</span>}
        {models.items && authzn && this.showList(models) }
        <Router history={hist}>
          <div>
            <PrivateRoute path={`${url}/:id`} component={detailsPage} />
          </div>
        </Router>
      </div>
    )
  }

/*  handleDeleteModel(id) {
    console.log('Deleting Permission with id: ', id)
    //return (e) => this.props.dispatch(actions.delete(id))
    this.props.dispatch(actions.delete(id))
    this.props.dispatch(actions.getAll()) // get list after deletion of a model
  } */
  showList(models){
    //const { authzn } = this.props
    return <Table>
      <thead>
        <tr>
          <th>#</th>
          <th>Operations</th>
          <th>Resource</th>
          <th>Conditions</th>
          <th>Description</th>
          <th>Actions {this.addLink()}</th>
        </tr>
      </thead>
      <tbody>
        {models.items.map((model, index) =>
          <tr key={model.id}>
            <th scope="row">{index+1}</th>
            <td>{model.operations}</td>
            <td>{model.resource}</td>
            <td>{model.condition}</td>
            <td>{model.description}</td>
            { this.showActions(model) }
          </tr>)}
      </tbody>
    </Table>
  }

  addLink() {
    const { authzn } = this.props
    return authzn.allowsAdd ?
      <Link to={{ pathname: `${url}/0`, state: this.newModel() }} title="Add"><MdAdd/></Link> :
      ''
  }

  newModel() {
    return {
      model: {
        id: 0,
        condition:'',
        description: ''
      }
    }
  }

  showActions(model) {
    const { authzn } = this.props
    return <td>
            <Link
              to={{ pathname: `${url}/${model.id}`, state:{model: model} }}
              title={authzn.allowsEdit?"Edit":"View"}
            >{authzn.allowsEdit?<MdEdit/>:<MdVisibility/>}</Link>
            <Button
              color="link"
              title="Delete"
              onClick={() => this.handleDeleteModel(model.id)}
              hidden={!authzn.allowsDelete}
            ><MdDelete color="red"/></Button>
          </td>
  }

  handleDeleteModel(id) {
    if( window.confirm('Are you sure?') ) {
      console.log('Delete confirmed')
      //this.props.dispatch(actions.delete(id))
      //this.props.dispatch(actions.getAll()) // get list after deletion of a model
      this.props.delete(id)
      this.props.getAll()
    }
  }


}

function mapStateToProps(state) {
  const { permissions, alert, authorizations } = state
  // const { user } = authentication
  const authzn = authorizations[module]
  const trackHistory = true  // added for unit testing; snapshot to be precise
  return {
//    user,
    permissions,
    alert,
    authzn,
    trackHistory
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getAll: () => {
      dispatch(actions.getAll())
    },
    delete: (id) => {
      dispatch(actions.delete(id))
    }
  }
}

//const connectedPermissionsPage = connect(mapStateToProps)(PermissionsPage)
//export { connectedPermissionsPage as PermissionsPage }
export default connect(mapStateToProps, mapDispatchToProps)(Permissions)
