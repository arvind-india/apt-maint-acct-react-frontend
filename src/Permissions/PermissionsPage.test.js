// PermissionsPage.test.js

import React from 'react'
import { shallow } from 'enzyme'

import { Permissions } from './PermissionsPage'

const mockProps = {
  trackHistory: false,  // for unit testing history is NOT required as it has 'keys' that are dynamic
  permissions: {
    loading: false,
    error: false,
    items: [
      {
        id: 1,
        operations: 'R',
        resource: 'ACCOUNT',
        conditions: '',
        description:'Grands Read-only permission on ACCOUNT',
      },
      {
        id: 2,
        operations: 'RU',
        resource: 'ACCOUNT',
        conditions: '',
        description:'Grands Read and Edit permissions on ACCOUNT',
      },
      {
        id: 3,
        operations: 'CRUD',
        resource: 'ACCOUNT',
        conditions: '',
        description:'Grands All permissions on ACCOUNT',
      }
    ]
  },
  alert: {message: "testingAlert"},
  authzn: {
    allowsAdd: true,
    allowsEdit: true,
    allowsDelete: true,
    condition: ''
  },
  getAll: jest.fn(),
  delete: jest.fn()
}

const component = shallow(<Permissions {...mockProps} />)

describe('PermissionsPage params test', () => {
  it('should render PermissionsPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
})

describe('PermissionsPage rows test', () => {
  let numberOfMockPermissions = mockProps.permissions.items.length
  let tbody = component.find('tbody')
  let rows = tbody.find('tr')
  it('should have '+numberOfMockPermissions+' rows', () => {
    expect(rows.length).toEqual(numberOfMockPermissions)
  })

})
