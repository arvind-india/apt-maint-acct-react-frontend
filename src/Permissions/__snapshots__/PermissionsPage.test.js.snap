// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionsPage params test should render initial layout 1`] = `
Array [
  <div>
    <h3>
        Permissions List
    </h3>
    <FlashMessage
        color={undefined}
        delay={2100}
        text="testingAlert"
    />
    <Table
        responsiveTag="div"
        tag="table"
    >
        <thead>
            <tr>
                <th>
                    #
                </th>
                <th>
                    Operations
                </th>
                <th>
                    Resource
                </th>
                <th>
                    Conditions
                </th>
                <th>
                    Description
                </th>
                <th>
                    Actions 
                    <Link
                        replace={false}
                        title="Add"
                        to={
                            Object {
                                "pathname": "/permissions/0",
                                "state": Object {
                                  "model": Object {
                                    "condition": "",
                                    "description": "",
                                    "id": 0,
                                  },
                                },
                              }
                        }
                    >
                        <MdAdd />
                    </Link>
                </th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <th
                    scope="row"
                >
                    1
                </th>
                <td>
                    R
                </td>
                <td>
                    ACCOUNT
                </td>
                <td />
                <td>
                    Grands Read-only permission on ACCOUNT
                </td>
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/permissions/1",
                                "state": Object {
                                  "model": Object {
                                    "conditions": "",
                                    "description": "Grands Read-only permission on ACCOUNT",
                                    "id": 1,
                                    "operations": "R",
                                    "resource": "ACCOUNT",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <th
                    scope="row"
                >
                    2
                </th>
                <td>
                    RU
                </td>
                <td>
                    ACCOUNT
                </td>
                <td />
                <td>
                    Grands Read and Edit permissions on ACCOUNT
                </td>
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/permissions/2",
                                "state": Object {
                                  "model": Object {
                                    "conditions": "",
                                    "description": "Grands Read and Edit permissions on ACCOUNT",
                                    "id": 2,
                                    "operations": "RU",
                                    "resource": "ACCOUNT",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <th
                    scope="row"
                >
                    3
                </th>
                <td>
                    CRUD
                </td>
                <td>
                    ACCOUNT
                </td>
                <td />
                <td>
                    Grands All permissions on ACCOUNT
                </td>
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/permissions/3",
                                "state": Object {
                                  "model": Object {
                                    "conditions": "",
                                    "description": "Grands All permissions on ACCOUNT",
                                    "id": 3,
                                    "operations": "CRUD",
                                    "resource": "ACCOUNT",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
        </tbody>
    </Table>
    <Router
        history={Object {}}
    >
        <div>
            <PrivateRoute
                component={[Function]}
                path="/permissions/:id"
            />
        </div>
    </Router>
</div>,
]
`;
